go/enclave/components/rollup_compression.go

```diff
@@ -0,0 +1,506 @@
+	incompleteBatches, err := rc.createIncompleteBatches(calldataRollupHeader, transactionsPerBatch, rollup.Header.L1Proof)
+	if err != nil {
+		return err
+	}
+
+	// 2. execute each batch to be able to calculate the hash which is necessary for the next batch as it is the parent.
+	err = rc.executeAndSaveIncompleteBatches(calldataRollupHeader, incompleteBatches)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// the main logic that goes from a list of batches to the rollup header
+func (rc *RollupCompression) createRollupHeader(batches []*core.Batch) (*common.CalldataRollupHeader, error) {
+	reorgs := make([]*common.BatchHeader, len(batches))
+
+	deltaTimes := make([]*big.Int, len(batches))
+	startTime := batches[0].Header.Time
+	prev := startTime
+
+	l1HeightDeltas := make([]*big.Int, len(batches))
+	var prevL1Height *big.Int
+
+	batchHashes := make([]common.L2BatchHash, len(batches))
+	batchHeaders := make([]*common.BatchHeader, len(batches))
+
+	isReorg := false
+	for i, batch := range batches {
+		rc.logger.Info("Add batch to rollup", log.BatchSeqNoKey, batch.SeqNo(), log.BatchHeightKey, batch.Number(), log.BatchHashKey, batch.Hash())
+		// determine whether the batch is canonical
+		can, err := rc.storage.FetchBatchByHeight(batch.NumberU64())
+		if err != nil {
+			return nil, err
+		}
+		if can.Hash() != batch.Hash() {
+			// if the canonical batch of the same height is different from the current batch
+			// then add the entire header to a "reorgs" array
+			reorgs[i] = batch.Header
+			isReorg = true
+			rc.logger.Info("Reorg", "pos", i)
+		} else {
+			reorgs[i] = nil
+		}
+		batchHashes[i] = batch.Hash()
+		batchHeaders[i] = batch.Header
+
+		deltaTimes[i] = big.NewInt(int64(batch.Header.Time - prev))
+		prev = batch.Header.Time
+
+		// since this is the sequencer, it must have all the blocks, because it created the batches in the first place
+		block, err := rc.storage.FetchBlock(batch.Header.L1Proof)
+		if err != nil {
+			return nil, err
+		}
+
+		// the first element is the actual height
+		if i == 0 {
+			l1HeightDeltas[i] = block.Number()
+		} else {
+			l1HeightDeltas[i] = big.NewInt(block.Number().Int64() - prevL1Height.Int64())
+		}
+		prevL1Height = block.Number()
+	}
+
+	l1DeltasBA := make([][]byte, len(l1HeightDeltas))
+	for i, delta := range l1HeightDeltas {
+		v, err := delta.GobEncode()
+		if err != nil {
+			return nil, err
+		}
+		l1DeltasBA[i] = v
+	}
+
+	timeDeltasBA := make([][]byte, len(deltaTimes))
+	for i, delta := range deltaTimes {
+		v, err := delta.GobEncode()
+		if err != nil {
+			return nil, err
+		}
+		timeDeltasBA[i] = v
+	}
+
+	reorgsBA, err := transformToByteArray(reorgs)
+	if err != nil {
+		return nil, err
+	}
+	// optimisation in case there is no reorg header
+	if !isReorg {
+		reorgsBA = nil
+	}
+
+	// get the first canonical batch
+	firstCanonBatchHeight := batches[0].Number()
+	firstCanonParentHash := batches[0].Header.ParentHash
+	for i, reorg := range reorgs {
```

Comment:
Took me a while to figure out this loop, maybe worth a little comment to say it's looping to find the first batch with no reorg or something.

Also, weird edge case to think about, if we somehow had a rollup that was only reorgs, would it brick the network because the first canon data wouldn't align with other rollups?